### PR TITLE
feat(authz): introduce generic scope-based authorization layer

### PR DESCRIPTION
**Pull request by sthanikan2000**
## Summary

Introduces `internal/authz` — a generic, scope-based HTTP authorization layer that complements `internal/auth`. Authorization is driven by the OAuth2 scopes carried on the access token. The package is fully decoupled from the authentication implementation by design: it defines its own minimal `Principal` interface and receives the principal through an injected `Extractor`.

Closes #561

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (no functional changes)

## Changes Made

- **`internal/authz/principal.go`** — `Principal` interface (`Subject()`, `Roles()`, `Scopes()`), `Extractor` type, `ErrUnauthenticated`/`ErrForbidden` sentinels, and `HasScope`/`HasAnyScope`/`HasAllScopes` helpers for service-layer checks.
- **`internal/authz/authorizer.go`** — `Authorizer` with `New(extract Extractor) (*Authorizer, error)`, `Principal(ctx)`, and `RequireScope`/`RequireAnyScope`/`RequireAllScopes` HTTP middleware (401 unauthenticated · 403 missing scope, house-style JSON errors, `slog` on 403).
- **`internal/authz/doc.go`** — package documentation.
- **`internal/authz/README.md`** — design principles, full API reference, composition-root wiring example, and service-layer usage pattern.

## Architecture

The package imports **nothing from `internal/auth`** or any other internal package (stdlib only). The authn layer's `*auth.AuthContext` satisfies `authz.Principal` *structurally* — the composition root wires `auth.GetAuthContext` into the `Extractor`, keeping the two packages independent.

```go
// Composition root (app bootstrap — not in this PR)
authzr, err := authz.New(func(ctx context.Context) (authz.Principal, bool) {
    ac := auth.GetAuthContext(ctx)
    if ac == nil || ac.Type() == "" { return nil, false }
    return ac, true // *auth.AuthContext satisfies authz.Principal structurally
})

// Per route
mux.Handle("GET /api/v1/consignments",
    withAuth(authzr.RequireScope("nsw:consignment:read")(h)))
```

Scope strings are app-defined constants at the composition root — never in this generic package.

## Testing

- [x] I have tested this change locally
- [x] I have added unit tests for new functionality
- [x] I have tested edge cases
- [x] All existing tests pass

Unit tests (`authz_test.go`) use a local `fakePrincipal` struct (zero authn dependency) to cover:
- `HasScope`/`HasAnyScope`/`HasAllScopes` table cases
- `New(nil)` returns error; factory panics on empty/no-scope args catch wiring bugs at startup
- `RequireScope` enforcement: 401 unauthenticated, 403 missing scope, 200 passthrough — for both user-style (roles + scopes) and client-style (scopes only) principals
- `RequireAnyScope` / `RequireAllScopes` enforcement
- `Principal(ctx)` including nil-receiver safety

Decoupling verified: `go list -deps ./internal/authz | grep internal/auth` → empty (zero dependency on the authn package).

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have checked that there are no merge conflicts

## Related Issues

Closes #561

## Additional Notes

This PR is pure plumbing — no existing handler, route, or service is touched and there is no behavioural change. Wiring `RequireScope` into routes and migrating the `?role=` query-param gate in `consignment/router.go` (TODO on line 85) are follow-up steps.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/internal/authz/authorizer.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	return a.require(func(p Principal) bool { return HasScope(p, scope) }, scope)
+}
+
+// RequireAnyScope admits the request if the principal holds at least one of scopes.
+func (a *Authorizer) RequireAnyScope(scopes ...string) func(http.Handler) http.Handler {
+	if len(scopes) == 0 {
+		panic("authz: RequireAnyScope requires at least one scope")
+	}
+	return a.require(func(p Principal) bool { return HasAnyScope(p, scopes...) }, scopes...)
+}
```
[the remark this answers is not in the document]
Comment (sthanikan2000) [reply backend/internal/authz/authorizer.go:62]: Addressed

File: backend/internal/authz/authorizer.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+// RequireAllScopes admits the request only if the principal holds every one of scopes.
+func (a *Authorizer) RequireAllScopes(scopes ...string) func(http.Handler) http.Handler {
+	if len(scopes) == 0 {
+		panic("authz: RequireAllScopes requires at least one scope")
+	}
+	return a.require(func(p Principal) bool { return HasAllScopes(p, scopes...) }, scopes...)
+}
```
[the remark this answers is not in the document]
Comment (sthanikan2000) [reply backend/internal/authz/authorizer.go:75]: Addressed

File: backend/internal/authz/authz_test.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+					t.Fatal("expected panic when no scopes provided")
+				}
+			}()
+			if name == "any" {
+				a.RequireAnyScope()
+			} else {
+				a.RequireAllScopes()
+			}
+		})
+	}
```
[the remark this answers is not in the document]
Comment (sthanikan2000) [reply backend/internal/authz/authz_test.go:186]: Addressed

**Review by mushrafmim — APPROVED** [review]

**Review by Aravinda-HWK — APPROVED** [review]
